### PR TITLE
Add Github subscription via websub

### PR DESCRIPTION
**Pull request by Chuhaa**
## Purpose
Add github subscription via websub

## Goals
Resolves https://github.com/wso2-enterprise/choreo/issues/14033

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Test environment
2201.2.1

### CODE REVIEW

File: asyncapi/github/dispatcher_service.bal [hunk 1]
```diff
@@ -43,36 +39,24 @@ service class DispatcherService {
...
-            if (trimmedSecret.length() != computedDigest.length() || trimmedSecret !== computedDigest) {
-                // Validate secret with X-Hub-Signature-256 header for intent verification
-                log:printError("Signature verification failure");
-                http:Response response = new;
-                response.statusCode = http:STATUS_UNAUTHORIZED;
-                response.setPayload("Signature verification failure");
-                return response;
-            }
+    remote function onEventNotification(websub:ContentDistributionMessage event)
+    returns Acknowledgement|websub:SubscriptionDeletedError|error? {
```
Comment (RolandHewage) [inline asyncapi/github/dispatcher_service.bal:43]: Let's align the return statement with the opening bracket of the function.

File: asyncapi/github/dispatcher_service.bal [hunk 2]
```diff
@@ -43,36 +39,24 @@ service class DispatcherService {
...
+        json payload = check event.content.ensureType(json);
+        map<string|string[]>? headers = event.headers;
+        string eventName = "";
+        if headers is map<string|string[]> {
+            string|string[] events = headers.get("X-Github-Event");
+            eventName = events is string ? events : events[0];
         }
         GenericDataType|error genericDataType = payload.cloneWithType(GenericDataType);
         if (genericDataType is error) {
```
Comment (RolandHewage) [inline asyncapi/github/dispatcher_service.bal:52]: Let's remove the opening & closing brackets here. It's a Ballerina antipattern

File: asyncapi/github/dispatcher_service.bal [hunk 3]
```diff
@@ -43,36 +39,24 @@ service class DispatcherService {
...
             log:printError("Unsupported Event Type : " + eventName);
             return genericDataType;
         } else {
             check self.matchRemoteFunc(genericDataType, eventName);
-            http:Response response = new;
-            response.statusCode = http:STATUS_OK;
-            response.setPayload("Event acknoledged successfully");
-            return response;
+            return {
+                body: {"message": "Event acknoledged successfully"}
```
Comment (RolandHewage) [inline asyncapi/github/dispatcher_service.bal:58]: spelling mistake here.
Suggestion: Event acknowledged

File: asyncapi/github/listener.bal [hunk 4]
```diff
@@ -15,25 +16,41 @@
...
+        self.websubListener = check new (listenOn);
+        self.dispatcherService = new DispatcherService();
     }
 
     public isolated function attach(GenericServiceType serviceRef, () attachPoint) returns @tainted error? {
         string serviceTypeStr = self.getServiceTypeStr(serviceRef);
         check self.dispatcherService.addServiceRef(serviceTypeStr, serviceRef);
+
+        websub:SubscriberServiceConfiguration? configuration = retrieveSubscriberServiceAnnotations(serviceRef);
+        if (configuration is websub:SubscriberServiceConfiguration) {
```
Comment (RolandHewage) [inline asyncapi/github/listener.bal:42]: Let's remove the opening & closing brackets here and in the following if blocks. It's a Ballerina anti pattern.

File: asyncapi/github/listener.bal [hunk 5]
```diff
@@ -15,25 +16,41 @@
...
+            self.config = <websub:SubscriberServiceConfiguration>configuration;
+            if (self.config?.httpConfig is http:ClientConfiguration) {
+                http:ClientConfiguration clientConfig = <http:ClientConfiguration>self.config?.httpConfig;
```
Comment (RolandHewage) [inline asyncapi/github/listener.bal:45]: Let's try to use ensureType() instead of manual casting wherever possible.

**Review by RolandHewage — COMMENTED** [review]
Added few comments.

File: asyncapi/github/listener.bal (hunk 5 above)
Comment (Chuhaa) [reply asyncapi/github/listener.bal]: Updated in https://github.com/ballerina-platform/asyncapi-triggers/commit/e0c56638e7d25694ff986c19c894b01fc776abb8

File: asyncapi/github/listener.bal (hunk 4 above)
Comment (Chuhaa) [reply asyncapi/github/listener.bal]: Fixed in https://github.com/ballerina-platform/asyncapi-triggers/commit/e0c56638e7d25694ff986c19c894b01fc776abb8

File: asyncapi/github/dispatcher_service.bal (hunk 2 above)
Comment (Chuhaa) [reply asyncapi/github/dispatcher_service.bal]: Updated in https://github.com/ballerina-platform/asyncapi-triggers/commit/e0c56638e7d25694ff986c19c894b01fc776abb8

File: asyncapi/github/dispatcher_service.bal (hunk 1 above)
Comment (Chuhaa) [reply asyncapi/github/dispatcher_service.bal]: Updated in https://github.com/ballerina-platform/asyncapi-triggers/commit/e0c56638e7d25694ff986c19c894b01fc776abb8

File: asyncapi/github/dispatcher_service.bal (hunk 3 above)
Comment (Chuhaa) [reply asyncapi/github/dispatcher_service.bal]: Fixed in https://github.com/ballerina-platform/asyncapi-triggers/commit/e0c56638e7d25694ff986c19c894b01fc776abb8

**Review by RolandHewage — APPROVED** [review]

**Comment by Drifftr** [comment]
Since this involves an API change we have to update related samples with this PR. And need to do a major version bump with it for the trigger.

**Comment by NipunaRanasinghe** [comment]
Closing as per https://github.com/wso2-enterprise/choreo/issues/14033#issuecomment-2329033317

**Pull request closed without merging**